rfcs/0171-fetch-from-github.md

```diff
@@ -0,0 +1,114 @@
+
+```
+# previous sha256 is still valid
+$ nix-build -A nix-template.src --check
+checking outputs of '/nix/store/ib74331l6pl6f8s2hsakf68lhg2jsl5i-nix-template-0.1.4-src.drv'...
+
+trying https://github.com/jonringer/nix-template/archive/v0.1.4.tar.gz
+  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                 Dload  Upload   Total   Spent    Left  Speed
+100   130  100   130    0     0    425      0 --:--:-- --:--:-- --:--:--   426
+100 27955    0 27955    0     0  36653      0 --:--:-- --:--:-- --:--:--  311k
+unpacking source archive /build/v0.1.4.tar.gz
+/nix/store/lfbgmqvpq5365q5ivv6ccck7xg88syw5-nix-template-0.1.4-src
+
+# explicit commit hash example
+$ nix-build -A artyFX.src
+this derivation will be built:
+  /nix/store/ir4k3n5q7nmb2wh533pq1ma1cabyr8h7-openAV-ArtyFX-8c542627d9-src.drv
+building '/nix/store/ir4k3n5q7nmb2wh533pq1ma1cabyr8h7-openAV-ArtyFX-8c542627d9-src.drv'...
+
+trying https://github.com/openAVproductions/openAV-ArtyFX/archive/8c542627d936a01b1d97825e7f26a8e95633f7aa.tar.gz
+  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                 Dload  Upload   Total   Spent    Left  Speed
+100   173  100   173    0     0    754      0 --:--:-- --:--:-- --:--:--   755
+100  627k    0  627k    0     0   604k      0 --:--:--  0:00:01 --:--:-- 1014k
+unpacking source archive /build/8c542627d936a01b1d97825e7f26a8e95633f7aa.tar.gz
+/nix/store/dkvcfm90ckrlgmv89s8sr15vcidwlxhs-openAV-ArtyFX-8c542627d9-src
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- All derivations which don't pass a "name" parameter will need to be re-realized
+    - This will be a download-intensive one-time cost to realize the new FOD derivations.
+    - NAR hash should not need to be recomputed assuming it was deterministic and not stale.
+    - Cache should be minimally impacted as NARs are content addressable, thus deterministic sources should not contribute to cache bloat.
+    - Potential for sources which are no longer available to be broken.
+        - These can have their name manually set to "source" to perserve previous behavior.
+        - Ideally source availability would be remedied with more appropriate methods. E.g. being made available.
+- "Interchangeability" with other fetchers is diminished as the derivation name is different
+    - In practice, fetchFromGitHub is never used in this way. It is generally the only fetcher, so there is never another FOD to dedupilicate.
+- Out-of-tree repositories may get hash mismatch errors
+    - If the cause of the mismatch is staleness, this is good and working as intended
+    - If the cause is non-determinism, this is frustrating.
+- Some derivations assume "source" to be the name of sourceRoot
+    - This has been mitigated over two years within Nixpkgs
+    - Out-of-tree code may break if they assume "source" is the name
+        - Can be mitigated with release notes describing the issue
```

Comment:
All these "drawbacks" are caused by the assumption of `src.name == "source"` being broken.

However, we have already deprecated such assumption as early as Nixpkgs 21.11. Nixpkgs Manual now requires `sourceRoot` to start with `"${src.name}"` instead of `"source"` when `src` is constructed with by `fetchgit`-based fetchers, and tree-wide conversions (#245388, #247977, #248528, #248683, #294334) has been merged.

`fetchFromGitHub` and other output-as-a-directory fetchers can still be used interchangeably if we stick to `${src.name}` instead of `"source"`.

In my opinion, we do not provide [bug-level compatibility](https://github.com/pkgconf/pkgconf#compatibility-with-pkg-config) to packages failing to follow already-stablized specifications, in-tree or out-of-tree.